source/common/upstream/original_dst_cluster.cc

```diff
@@ -47,22 +61,73 @@ HostConstSharedPtr OriginalDstCluster::LoadBalancer::chooseHost(LoadBalancerCont
       if (dst_ip) {
         Network::Address::InstanceConstSharedPtr host_ip_port(
             Network::Utility::copyInternetAddressAndPort(*dst_ip));
+        Network::TransportSocketOptionsSharedPtr transport_socket_options;
+
+        // Create an override transport socket options that automatically provides both SNI as well
+        // as SAN verification for the new host if the cluster has been configured with TLS.
+        if (parent_->implements_secure_transport_) {
+          absl::string_view sni;
+
+          // First check if have requested server name in the downstream connection
+          // This may be set by a TLS inspector listener socket
+          const Network::Connection* downstream_connection = context->downstreamConnection();
+          if (downstream_connection) {
+            sni = downstream_connection->requestedServerName();
+          }
+
+          if (sni.length() > 0) {
+            ENVOY_LOG(debug, "Using SNI from downstream connection: {}.", sni);
+          } else {
+            // If no SNI from downstream connection, check if have HTTP Host in downstream headers.
+            const Http::HeaderMap* downstream_headers = context->downstreamHeaders();
+            if (downstream_headers) {
+              auto host_header = downstream_headers->get(Http::Headers::get().Host);
+              if (host_header != nullptr) {
```

Comment:
Please don't copy this logic around more, see #9026 , this is blocking #8889 and #9139.